Implement playwright end-to-end browser-based tests

NSeydoux opened the pull request:
This adds playwright tests on top of the test app built in #229 . Some updates to the test app are introduced to better suit playwright: for instance, some elements are now not present in the DOM if they have no effect, instead of being present by default and handling internally whether clicks should have effects or not. This way, playwright can rely on selectors to find the elements, instead of relying on flaky timeouts.

There are currently only two tests: 
- One verifying that the socket connection works, and that connect/closed events are handled appropriately$
- One verifying that Pod events are sent in the websocket.

# Checklist

- [X] All acceptance criteria are met.
- [ ] Relevant documentation, if any, has been written/updated.
- [ ] The changelog has been updated, if applicable.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

NSeydoux committed:
Delete testcafe tests

NSeydoux committed:
Replace testcafe dependencies by playwright

NSeydoux committed:
WIP: Playwright tests

The test environment being down, end-to-end test development is paused for the moment.

NSeydoux committed:
Run playwright from test scripts

NSeydoux committed:
Ignore playwright test results

NSeydoux committed:
Only render button when websocket can be connected

Otherwise, playwright clicks before it has any effect.

NSeydoux committed:
Only display socket status after connection

Prevents from having an arbitrary await for react to render the state: instead, the element only appears in the DOM when ready

NSeydoux committed:
Add test checking websocket events

This adds a test creating and deleting resources on the Pod, and verifying that associated events are sent in the Websocket.

NSeydoux committed:
Simplify socket status display condition

NSeydoux committed:
Validate testing environment

NSeydoux committed:
Fix linting

ThisIsMissEm commented on e2e/browser/src/components/notifications/index.tsx:
```diff
@@ -21,13 +21,127 @@ const MessageList = (props: { messages: Array<any> }) => {
   );
 };
 
+const WebSocketButtons = ({ socket }: { socket?: WebsocketNotification }) => {
```
better to pass the handlers in here, rather than the socket itself, as the socket is a pretty complex object which may cause state changes.

NSeydoux replied:
1ffd2f6581286fd5ef9541ef04f6a5343a77b93d

ThisIsMissEm commented on e2e/browser/src/components/notifications/index.tsx:
```diff
@@ -21,13 +21,127 @@ const MessageList = (props: { messages: Array<any> }) => {
+          e.preventDefault();
+          await socket.connect();
+        }}
+        data-testid="connectSocket"
```
You could also have this change from a connect button to a disconnect button depending on connection state.

NSeydoux replied:
That's a good idea, I've updated the app slightly to that effect.

ThisIsMissEm commented on e2e/browser/src/components/notifications/index.tsx:
```diff
@@ -44,15 +158,15 @@ export default function Notifications() {
         new WebsocketNotification(parentContainerUrl, {
           fetch: getDefaultSession().fetch,
-          gateway: "https://notification.inrupt.com",
+          gateway: notificationGateway,
```
isn't this automatically discovered based on the pod host via the well-known?

NSeydoux replied:
That's correct, I was hung up on an old bug that has been long resolved. Removed in 8c85c8bcffd2cab41de1c9848eb81fbe5ae97f26

ThisIsMissEm commented on e2e/browser/test/README.md:
```diff
@@ -1,4 +1,4 @@
-# End-to-end tests for solid-client-notifications in the browser
+# End-to-end tests for solid-client in the browser
```
This change looks incorrect?

NSeydoux replied:
Indeed f6722f74d955263a111480c64bff3bd77eb501e0

ThisIsMissEm commented on e2e/browser/test/README.md:
```diff
@@ -7,89 +7,53 @@ description of our setup.
-signatures from `-helpers` without executing the code that sets up the
-listeners. These helpers are then stored on the global variable `E2eHelpers` in
-`end-to-end-test.ts`. Why this is needed will be explained in the next section.
+One thing to note is that its dependency on solid-client-access-grants should be
```
```suggestion
One thing to note is that its dependency on `solid-client-notifications` should be
```

NSeydoux replied:
f6722f74d955263a111480c64bff3bd77eb501e0

ThisIsMissEm commented on e2e/browser/test/e2e.playwright.ts:
```diff
@@ -0,0 +1,136 @@
+const { login, password, notificationGateway } = getTestingEnvironmentBrowser();
+
+test.skip("connecting a websocket and disconnecting it", async ({ page }) => {
+  let websocket: PlayWrigtWebSocket;
```
```suggestion
  let websocket: PlayWrightWebSocket;
```

NSeydoux replied:
f6722f74d955263a111480c64bff3bd77eb501e0

ThisIsMissEm commented on e2e/browser/test/e2e.playwright.ts:
```diff
@@ -0,0 +1,136 @@
+import {
+  test,
+  expect,
+  WebSocket as PlayWrigtWebSocket,
```
```suggestion
  WebSocket as PlayWrightWebSocket,
```

NSeydoux replied:
f6722f74d955263a111480c64bff3bd77eb501e0

ThisIsMissEm commented on e2e/browser/test/e2e.playwright.ts:
```diff
@@ -0,0 +1,136 @@
+    // The resource creation should trigger a websocket frame
+    websocket.waitForEvent("framereceived"),
+    page.click("button[data-testid=createContainer]"),
+  ]);
```
This is a kind of weird looking way of doing things, especially as all these promises will potentially execute in parallel; You could perhaps do a `websocket.on("framereceived", (payload) => events.push(payload));` and then assert that the last item is the event you expected after clicking the button to create the container?

NSeydoux replied:
c6aa1edf670687ffab221f6e2e2b2ac6446874f9

ThisIsMissEm commented on e2e/browser/test/e2e.playwright.ts:
```diff
@@ -0,0 +1,136 @@
+  await Promise.all([
+    page.click("button[data-testid=deleteContainer]"),
+    page.click("button[data-testid=disconnectSocket]"),
+  ]);
```
Then you could also assert down here that you've actually received the deleteContainer event, before you disconnect the socket.

NSeydoux replied:
c6aa1edf670687ffab221f6e2e2b2ac6446874f9

ThisIsMissEm commented on e2e/browser/test/pageModels/index.ts:
```diff
@@ -1,67 +1,48 @@
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
```
I think it's a copy paste artifact, but this is the incorrect license header.

NSeydoux replied:
88f47b24111cd5c3a21326994293cbdd7a6bf894

ThisIsMissEm commented on e2e/browser/test/playwright.config.ts:
```diff
@@ -0,0 +1,74 @@
+      name: "Firefox",
+      use: {
+        browserName: "firefox",
+        userAgent: `Browser-based solid-client-access-grant end-to-end tests running ${
```
```suggestion
        userAgent: `Browser-based solid-client-notifications end-to-end tests running ${
```

ThisIsMissEm reviewed (changes requested):
Overall looking like good progress, but I think there's a bunch of things that need to be fixed or changed.

NSeydoux committed:
Pass handlers instead of more complex socket

NSeydoux committed:
Pass down handler instead of setters

NSeydoux committed:
Remove notification gateway management

NSeydoux commented on e2e/browser/src/components/notifications/index.tsx:
```diff
@@ -21,13 +21,127 @@ const MessageList = (props: { messages: Array<any> }) => {
+    <button
+      onClick={async (e) => {
+        e.preventDefault();
+        setChildContainerUrl(
```
94c88acb7165ce48755c1b4a76676aea304a1cf1

NSeydoux replied:
Do you have resources in mind explaining why this is an anti-pattern, or alternatively would you mind having a chat sometime to walk me through the issue so that I understand it better ?

NSeydoux committed:
Fix the test resdme

NSeydoux committed:
Fix typo

NSeydoux committed:
Assert on the received event

NSeydoux committed:
Fix licence headers

NSeydoux committed:
Fix e2e workflow file

NSeydoux committed:
Unskip test

NSeydoux committed:
Fix path in breowser test build script

NSeydoux committed:
Add missing variables to browser-based tests

NSeydoux committed:
Add assertions on the notification messages

Test if mesages are actually coming through when events happen in the container.

NSeydoux committed:
Set env for e2e in CI

NSeydoux committed:
Only require client credential for node tests

NSeydoux committed:
Run browser tests headless

NSeydoux committed:
Fix trace file path

NSeydoux committed:
Skip tests until websocket bug is resolved

NSeydoux commented:
Unfortunately for the time being a bug makes these tests flaky. I suggest we merge this with the tests skipped until the bug is resolved so that this PR doesn't linger around in the meantime.
